### ISSUE TITLE
false positive `dead_code` "never constructed" warning on 1.80.0 for a `struct` where the constructor takes `&'static self`

### ISSUE THREAD

**Issue by chrisnc** [rust-lang/rust#128272]
### Code

```Rust
/* XXX: changing Foo::init to take self: &'static Foo instead of
 * &'static self silences the dead_code warning. */
pub struct Foo {
    x: i32,
}

impl Foo {
    #[must_use]
    //pub const fn init(self: &'static Foo, x: i32) -> Foo {
    pub const fn init(&'static self, x: i32) -> Foo {
        Foo { x }
    }

    pub fn foo(&self) -> i32 {
        self.x
    }
}

#[cfg(test)]
mod test {
    use crate::Foo;

    #[test]
    fn test() {
        static FOO: Foo = FOO.init(7);
        assert_eq!(FOO.foo(), FOO.x);
    }
}
```

### Current output

```Shell
warning: struct `Foo` is never constructed
 --> src/lib.rs:3:12
  |
3 | pub struct Foo {
  |            ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `playground` (lib) generated 1 warning
```

### Desired output

No output.

### Rationale and extra context

This problem does not occur with 1.79.0. https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=ad761b02ac5d86f28629e7fa2d6a1936

This is a reduced example from real code where objects are constructed in such a way that they need static references to fields within the same object (and those objects themselves will be static). It's clearly possible to construct such an object, so the dead_code warning is not accurate.

### Other cases

Changing the parameter to be `self: &'static Foo` instead silences the warning.

### Rust Version

```Shell
$ rustc --version --verbose
rustc 1.80.0 (051478957 2024-07-21)
binary: rustc
commit-hash: 051478957371ee0084a7c0913941d2a8c4757bb9
commit-date: 2024-07-21
host: aarch64-apple-darwin
release: 1.80.0
LLVM version: 18.1.7
```

**Comment by tgross35** [comment]
How do you wind up with the `&'static self` in the original code, is it via FFI? If so, https://github.com/rust-lang/rust/issues/126169 / https://github.com/rust-lang/rust/pull/127104 may be relevant.

**Comment by chrisnc** [comment]
The example I pasted shows how the `&'static self` is obtained: `static FOO: Foo = FOO.init(7);`
This part matches exactly what the original code was doing. I'm not using any `repr` attribute in this example so I'm not sure it's related to those issues. The warning should fire for this struct if the `init` function didn't exist, but `init` does provide a valid way to construct `Foo`.

**Comment by tgross35** [comment]
Completely missed the example, thanks. This is a pretty weird case considering `self` is the parameter that doesn't get used.

**Comment by chrisnc** [comment]
In the real code the `&'static self` does get used of course, but as I was minimizing it I found that it wasn't necessary. The `'static` itself is not necessary to see the warning either, but in this case I think the only way to call `init` in safe rust is with a `static`, so I included it for clarity.

**Comment by chrisnc** [comment]
Some bisecting shows that this is caused by #125572.

**Comment by theemathas** [comment]
Minimal reproduction: a library crate containing the following code

```rust
pub struct Foo(&'static Foo);  // The field is included for demonstration of how this could be useful

impl Foo {
    pub const fn chain(&'static self) -> Foo {
        Foo(self)
    }
}

// An external crate could construct Foo by writing:
// static X: Foo = X.chain();
```

<details>
<summary>Compiler output on nightly 2024-07-12</summary>

```
   Compiling playground v0.0.1 (/playground)
warning: struct `Foo` is never constructed
 --> src/lib.rs:1:12
  |
1 | pub struct Foo(&'static Foo);  // The field is included for demonstration of how this could be useful
  |            ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `playground` (lib) generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.47s
```

</details>

**Comment by gurry** [comment]
A self-contained reproducer:
```rust
mod a {
    pub struct Foo(&'static Foo);  // The field is included for demonstration of how this could be useful
    
    impl Foo {
        pub const fn chain(&'static self) -> Foo {
            Foo(self)
        }
    }
}

use a::Foo;

fn main() {
    static _X: Foo = _X.chain();
}
```

**Comment by mu001999** [comment]
> In the real code the `&'static self` does get used of course

@chrisnc Hi, I'm curious about what do you need an uninitialized self (`&'static self`) for?

**Comment by chrisnc** [comment]
@mu001999 the original comment gives some explanation:

> This is a reduced example from real code where objects are constructed in such a way that they need static references to fields within the same object (and those objects themselves will be static).

The reference is not uninitialized (nor is the object, it's statically initialized).

**Comment by mu001999** [comment]
@chrisnc oh sorry, I missed it. i think i understand ;)

**Comment by RalfJung** [comment]
> The reference is not uninitialized (nor is the object, it's statically initialized, self-referentially).

When `init` runs, the memory `self` points to is actually uninitialized. Turns out that's a soundness bug so we'll probably have to reject such code or make it unsafe... see https://github.com/rust-lang/rust/issues/143047.

**Comment by chrisnc** [comment]
@RalfJung thanks for the heads up. As a data point for what to do about this, I hit the original bug here in real code that used this trick to statically initialize self-referential objects that were used by an FFI. That code has since migrated[^1] to use raw pointers and `&raw mut` once that became legal to do in `const` in Rust 1.83.0 (#129195), though it still takes a reference to a `static` `UnsafeCell`, which it then constructs by first calling `get` on the `UnsafeCell`, so maybe that is still subtly invalid too? Maybe there is some general transformation of this into equivalent raw pointer operations that would suffice for other code in the wild that does similar shenanigans, but maybe not. 

[^1]: https://git.rtng.org/rt/rt/commit/1a349f6eea269beea5395f8bf71913629ef97b06#diff-d65a61ea6076a114c1230c62fc30b9194ce51e48
